core/src/main/java/io/grpc/internal/ManagedChannelImplBuilder.java

```diff
@@ -675,26 +694,45 @@ public ManagedChannelImplBuilder addMetricSink(MetricSink metricSink) {
 
   @Override
   public ManagedChannel build() {
+    ClientTransportFactory clientTransportFactory =
+        clientTransportFactoryBuilder.buildClientTransportFactory();
+    ResolvedNameResolver resolvedResolver = getNameResolverProvider(
+        target, nameResolverRegistry, clientTransportFactory.getSupportedSocketAddressTypes());
     return new ManagedChannelOrphanWrapper(new ManagedChannelImpl(
         this,
-        clientTransportFactoryBuilder.buildClientTransportFactory(),
+        clientTransportFactory,
+        resolvedResolver.targetUri,
+        resolvedResolver.provider,
         new ExponentialBackoffPolicy.Provider(),
         SharedResourcePool.forResource(GrpcUtil.SHARED_CHANNEL_EXECUTOR),
         GrpcUtil.STOPWATCH_SUPPLIER,
-        getEffectiveInterceptors(),
+        getEffectiveInterceptors(resolvedResolver.targetUri.toString()),
         TimeProvider.SYSTEM_TIME_PROVIDER));
   }
 
   // Temporarily disable retry when stats or tracing is enabled to avoid breakage, until we know
   // what should be the desired behavior for retry + stats/tracing.
   // TODO(zdapeng): FIX IT
   @VisibleForTesting
-  List<ClientInterceptor> getEffectiveInterceptors() {
+  List<ClientInterceptor> getEffectiveInterceptors(String computedTarget) {
+    List<ClientInterceptor> effectiveInterceptors = new ArrayList<>(this.interceptors);
+    for (int i = 0; i < effectiveInterceptors.size(); i++) {
```

Comment:
I think it would be more obvious if you looped over interceptors building up effectiveInterceptors.